### PR TITLE
Ingest events into DB

### PR DESCRIPTION
**Pull request by psheth9**
Fixes: https://github.com/stellar/soroban-rpc/issues/189

This diff mainly covers the write path (ingestion) of events, follow up PR refactors getEvents in order to fetch it from DB

Changes:

- Introduce new event table
- Parse events from LCM and ingest in DB
- Tests 

Design Doc [here](https://docs.google.com/document/d/1Q5AJvszwFmwl3NzHh7PJIMj4mhN3HgHRn_7IOQGeMWg/edit)
Parent Epic [here](https://github.com/stellar/soroban-rpc/issues/115)

### CODE REVIEW

**Comment by 2opremio** [comment]
Please do not merge until there is a solution for data migration

**Comment by psheth9** [comment]
> Please do not merge until there is a solution for data migration

Not going to merge to `main` but to `events-db-backend` !! @2opremio

**Comment by 2opremio** [comment]
Did you write migration code for the new schema? Running instances with the in-memory datasctructure should work when running this code.

**Comment by 2opremio** [comment]
You also need to remove the in-memory prefill from daemon.go

**Comment by psheth9** [comment]
@2opremio Yes I will be adding the migration code and going to remove the entire in-memory storage of events in follow up PR. Not planning to do that in this PR bc need to refactor `getEvents` [to fetch it from DB instead of in-memory] 

Could you approve this PR if DB schema and ingestion looks good :) ? Updated the details of follow up refactoring task [here](https://github.com/stellar/soroban-rpc/issues/190) 

Note: This won't merge changes to `main` but only to` events-db-backend`

File: cmd/soroban-rpc/internal/db/event.go
Comment (aditya1702) [inline cmd/soroban-rpc/internal/db/event.go:18]: I think the interface should not be defined in the same file as its implementation: https://go.dev/wiki/CodeReviewComments#interfaces

Would like to get a recommendation from @tamirms about this.

File: cmd/soroban-rpc/internal/db/event.go
Comment (aditya1702) [inline cmd/soroban-rpc/internal/db/event.go:67]: Why are we skipping if the transaction is not successful? There can be diagnostic events for both successful and failed txns right?

@2opremio @tamirms

File: cmd/soroban-rpc/internal/db/event_test.go
Comment (2opremio) [inline cmd/soroban-rpc/internal/db/event_test.go:166]: Can you please resolve this before merging?

File: cmd/soroban-rpc/internal/db/migrations/03_events.sql
Comment (2opremio) [inline cmd/soroban-rpc/internal/db/migrations/03_events.sql:9]: Nit: formatting

**Comment by 2opremio** [comment]
Can you please fix the tests and the TODO?

File: cmd/soroban-rpc/internal/db/event.go
Comment (psheth9) [reply cmd/soroban-rpc/internal/db/event.go:67]: Yeah this is the existing behavior. Can update it if it is not true anymore.

Comment (aditya1702) [reply cmd/soroban-rpc/internal/db/event.go:67]: Yeah I saw it is in current getEvents too. I wanted to confirm this is true since I saw diagnostic events in both types of txns.

File: cmd/soroban-rpc/internal/db/event_test.go
Comment (psheth9) [reply cmd/soroban-rpc/internal/db/event_test.go:166]: yes, actually ignore this TODO as `getEvents` is backed by memory in this diff, this will be resolved in this follow up PR https://github.com/stellar/soroban-rpc/pull/215 (WIP) @2opremio

**Comment by psheth9** [comment]
@2opremio  Fixed the tests PTAL :) Next is coming up soon

**Comment by 2opremio** [comment]
Ah, the in-memory store still exists? I would much rather review finished, self-contained code with proper tests.

Can you remove the in-memory store and add end-to-end tests here?

**Comment by 2opremio** [comment]
I have finished what I was going , so I will get to review this on Monday. Sorry for the delay.

File: cmd/soroban-rpc/internal/db/event.go
Comment (2opremio) [reply cmd/soroban-rpc/internal/db/event.go:18]: You are right, but that's how it's done everywhere else in the db package

File: cmd/soroban-rpc/internal/db/event.go
Comment (2opremio) [inline cmd/soroban-rpc/internal/db/event.go:52]: Don't overwrite the error, just use `errors.Join()`

Comment (2opremio) [reply cmd/soroban-rpc/internal/db/event.go:52]: Also, note that err is shadowed down the line, so this may not be the `err` you want

File: cmd/soroban-rpc/internal/db/event.go
Comment (2opremio) [inline cmd/soroban-rpc/internal/db/event.go:71]: This shadows the `err`above, so the error handling in defer above won't do what you expect. You may want to use a named return parameter to make it cleaner.

**Comment by 2opremio** [comment]
The approach seems fine, but ... it's hard to tell without seeing the complexity of the querying side.

I would like to see a benchmark of querying and see how it compares with the in-memory store.

**Review by 2opremio — APPROVED** [review]

File: cmd/soroban-rpc/internal/db/event.go
Comment (2opremio) [inline cmd/soroban-rpc/internal/db/event.go:88]: @psheth9 I have just realized that the table already contains a ledger row, so the ledger info here is redundant.

Also:

1. if the op is always 0 (and we don't plan that to change) it shouldn't go into the DB.
2. the tx index isn't the same as the application order field?

Thus, unless there is a good reason for the opposite, I would suggest storing the tx index and event index instead of the id field.

**Pull request merged**
